pkg/build/gobuild.go

```diff
@@ -273,15 +284,45 @@ func build(ctx context.Context, ip string, platform v1.Platform, disableOptimiza
 		}
 	}
 
-	cmd.Env = append(defaultEnv, os.Environ()...)
+	args := make([]string, 0, 7)
+	args = append(args, "build")
+	if disableOptimizations {
+		// Disable optimizations (-N) and inlining (-l).
+		args = append(args, "-gcflags", "all=-N -l")
+	}
+	args = addGo113TrimPathFlag(args)
+
+	defaultEnv = append(defaultEnv, os.Environ()...)
+
+	tmpDir, err := ioutil.TempDir("", "ko")
+	if err != nil {
+		return "", err
+	}
+	if os.Getenv("KO_STABLE_OUTPUT") != "" {
+		hasher := md5.New() //nolint: gosec // No strong cryptography needed.
+		hasher.Write([]byte(strings.Join(args, " ") + " " + strings.Join(defaultEnv, " ")))
+
+		tmpDir = filepath.Join(os.TempDir(), "ko", ip, hashInputs(args, defaultEnv))
+		if err := os.MkdirAll(tmpDir, os.ModePerm); err != nil {
+			return "", err
+		}
+	}
+	file := filepath.Join(tmpDir, "out")
+
+	args = append(args, "-o", file)
+	args = append(args, ip)
+	cmd := exec.CommandContext(ctx, "go", args...)
+	cmd.Env = defaultEnv
 
 	var output bytes.Buffer
 	cmd.Stderr = &output
 	cmd.Stdout = &output
 
 	log.Printf("Building %s for %s", ip, platformToString(platform))
 	if err := cmd.Run(); err != nil {
-		os.RemoveAll(tmpDir)
+		if os.Getenv("KO_STABLE_OUTPUT") == "" {
+			os.RemoveAll(tmpDir)
+		}
```

Comment:
Why cache if it is an error?

Reply:
Because we don't want to delete the other metadata in that directory, even though the `out` binary is bad (or never got built). I think I want to rework this directory structure a bit before we merge it.